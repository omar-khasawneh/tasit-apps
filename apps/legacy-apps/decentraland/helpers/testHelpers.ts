import { ESTATE, PARCEL } from "../constants/AssetTypes";
import { DONE, PENDING, MISSING } from "../constants/ActionStatus";
import { SUCCESSFUL } from "../constants/UserActionStatus";

export const parcel = {
  type: PARCEL,
  id: "0123456789",
  name: "Sample Parcel",
  img: "https://api.decentraland.org/v1/parcels/0/0/map.png",
};

export const estate = {
  type: ESTATE,
  id: "123",
  name: "Sample Estate",
  img: "https://api.decentraland.org/v1/estates/1/map.png",
};

export const estateWithoutName = {
  type: ESTATE,
  id: "321",
  name: "",
  img: "https://api.decentraland.org/v1/estates/1/map.png",
};

export const estateForSale = {
  id: "1",
  priceMana: "0",
  priceUSD: 0,
  asset: estate,
};

export const parcelForSale = {
  id: "2",
  priceMana: "0",
  priceUSD: 0,
  asset: parcel,
};

function getIdAsync(): Promise<string> {
  return new Promise((resolve, reject) => {
    process.nextTick(() =>
      resolve("0x1234567890123456789012345678901234567890")
    );
  });
}

export const anAction = {
  getId: async (): Promise<string> => {
    return await getIdAsync();
  },
};

export const accountCreationSteps = [
  {
    name: "Account created",
    status: DONE,
  },
  {
    name: "ETH",
    status: PENDING,
  },
  {
    name: "MANA",
    status: MISSING,
  },
];

export const parcelUserAction = {
  ["0x1234567890123456789012345678901234567890"]: {
    assetId: "0123456789",
    status: SUCCESSFUL,
  },
};

export const estateUserAction = {
  ["0x987654321098765432109876543210987654321"]: {
    assetId: "123",
    status: SUCCESSFUL,
  },
};

export const userActions = { ...parcelUserAction, ...estateUserAction };

export default {
  estateForSale,
  parcelForSale,
  parcel,
  estate,
  anAction,
  estateWithoutName,
  parcelUserAction,
  estateUserAction,
  userActions,
};
