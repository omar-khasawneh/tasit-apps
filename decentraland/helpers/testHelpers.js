import AssetTypes from "@constants/AssetTypes";
import { DONE, PENDING, MISSING } from "@constants/ActionStatus";
const { ESTATE, PARCEL } = AssetTypes;

export const parcel = {
  type: PARCEL,
  id: "0123456789",
  name: "Sample Parcel",
  img: "https://api.decentraland.org/v1/parcels/0/0/map.png",
};

export const estate = {
  type: ESTATE,
  id: "123",
  name: "Sample Estate",
  img: "https://api.decentraland.org/v1/estates/1/map.png",
};

export const estateWithoutName = {
  type: ESTATE,
  id: "321",
  name: "",
  img: "https://api.decentraland.org/v1/estates/1/map.png",
};

export const estateForSale = {
  id: "1",
  priceMana: "0",
  priceUSD: 0,
  asset: estate,
};

export const parcelForSale = {
  id: "2",
  priceMana: "0",
  priceUSD: 0,
  asset: parcel,
};

export const anAction = {
  getId: async () => "0x1234567890123456789012345678901234567890",
};

export const accountCreationSteps = [
  {
    name: "Account created",
    status: DONE,
  },
  {
    name: "ETH",
    status: PENDING,
  },
  {
    name: "MANA",
    status: MISSING,
  },
];

export default {
  estateForSale,
  parcelForSale,
  parcel,
  estate,
  anAction,
  estateWithoutName,
  accountCreationSteps,
};
