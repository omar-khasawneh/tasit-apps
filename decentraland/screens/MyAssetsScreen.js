import React from "react";
import { connect } from "react-redux";
import { removeFromMyAssetsList, appendToMyAssetsList } from "../redux/actions";
import PropTypes from "prop-types";
import MyAssetsList from "@presentational/MyAssetsList";
import MyAssetsListItem from "@presentational/MyAssetsListItem";
import { listsAreEqual, getContracts, logWarn } from "@helpers";
import { generateAssetFromId } from "@helpers/decentraland";
import DecentralandUtils from "tasit-sdk/dist/helpers/DecentralandUtils";
import MyAssetStatus from "@constants/MyAssetStatus";
const { BOUGHT } = MyAssetStatus;

export class MyAssetsScreen extends React.Component {
  componentDidMount = async () => {
    const {
      account,
      myAssets: assetsFromState,
      removeFromMyAssetsList,
      appendToMyAssetsList,
    } = this.props;
    if (account) {
      const { address } = account;

      // Note: If a stored asset doesn't have a status, assuming 'bought'.
      const boughtAssets = assetsFromState
        .map(asset => (!asset.status ? { ...asset, status: BOUGHT } : asset))
        .filter(asset => asset.status === BOUGHT);

      const assetsFromBlockchain = await this._getAssetsFromBlockchain(address);
      const shouldUpdate = !listsAreEqual(assetsFromBlockchain, boughtAssets);
      if (shouldUpdate) {
<<<<<<< HEAD
        this._logAssetsInconsistency(assetsFromBlockchain, boughtAssets);
=======
        // TODO: Add some UI indication that something unexpected happened
        logWarn(
          `Account's assets from blockchain aren't the same as the ones stored on the app.`
        );
>>>>>>> be85215bf97edb9c9ec4a201f2183f829b9f6a28
        removeFromMyAssetsList(boughtAssets);
        appendToMyAssetsList(assetsFromBlockchain);
      }
    }
  };

  _logAssetsInconsistency = (fromBlockchain, fromState) => {
    const fromBlockchainIds = fromBlockchain.map(asset => asset.id);
    const fromStateIds = fromState.map(asset => asset.id);

    const addedIds = fromBlockchainIds.filter(id => !fromStateIds.includes(id));
    const removedIds = fromStateIds.filter(
      id => !fromBlockchainIds.includes(id)
    );

    // TODO: Add some UI indication that something unexpected happened
    logWarn(`Assets from blockchain aren't the same as the stored on the app.`);

    if (addedIds.length > 0)
      logWarn(`Some assets added to MyLandScreen. IDs: [${addedIds}]`);

    if (removedIds.length > 0)
      logWarn(`Some assets removed from MyLandScreen. IDs: [${removedIds}]`);
  };

  _getAssetsFromBlockchain = async address => {
    const listOfPromises = await this._getAssetsOf(address);
    const fromBlockchain = await Promise.all([...listOfPromises]);
    // Note: The UI shows land the user bought ordered by purchase date (desc)
    return fromBlockchain.reverse();
  };

  // Note: Returns a list of Promises
  _getAssetsOf = async address => {
    const decentralandUtils = new DecentralandUtils();
    const { getAssetsOf: getLandOf } = decentralandUtils;

    const contracts = getContracts();
    const { estateContract, landContract } = contracts;

    const listOfLand = await getLandOf(address);
    const assets = [];

    for (let land of listOfLand) {
      const generateAsset = async land => {
        const { id: assetId, nftAddress, transactionHash: actionId } = land;
        const status = BOUGHT;

        const asset = await generateAssetFromId(
          estateContract,
          landContract,
          assetId,
          nftAddress
        );

        return { ...asset, actionId, status };
      };

      const assetPromise = generateAsset(land);
      assets.push(assetPromise);
    }

    return assets;
  };

  _renderItem = ({ item: asset }) => {
    return <MyAssetsListItem asset={asset} />;
  };

  render() {
    const { myAssets } = this.props;

    return (
      <MyAssetsList myAssetsList={myAssets} renderItem={this._renderItem} />
    );
  }
}

MyAssetsScreen.propTypes = {
  myAssets: PropTypes.array.isRequired,
  account: PropTypes.object,
};

const mapStateToProps = state => {
  const { myAssets, accountInfo } = state;
  const { account } = accountInfo;
  const { list: myAssetsList } = myAssets;
  return { myAssets: myAssetsList, account };
};

const mapDispatchToProps = {
  removeFromMyAssetsList,
  appendToMyAssetsList,
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(MyAssetsScreen);
