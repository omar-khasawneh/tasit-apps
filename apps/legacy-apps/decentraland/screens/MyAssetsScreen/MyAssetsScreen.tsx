import React, { useEffect } from "react";
import { useDispatch, useSelector } from "react-redux";

import { GlobalState } from "../../types/GlobalState";

import {
  removeFromMyAssetsList,
  appendToMyAssetsList,
  addUserAction,
} from "../../redux/actions";
import MyAssetsList from "../../components/presentational/MyAssetsList";
import MyAssetsListItem from "../../components/presentational/MyAssetsListItem";
import { listsAreEqual, getContracts, logInfo, logWarn } from "../../helpers";
import { generateAssetFromId } from "../../helpers/decentraland";
import DecentralandUtils from "tasit-sdk/dist/helpers/DecentralandUtils";
import { SUCCESSFUL } from "../../constants/UserActionStatus";

interface AccountObject {
  address: string;
}

interface AssetObject {
  actionId: string;
  id: string;
}

interface UserActionObject {
  assetId: string;
  status: string;
}

interface AssetsAndActionsObject {
  assetsFromBlockchain: AssetObject[];
  actionsFromBlockchain: object;
}

interface SelectedState {
  myAssets: AssetObject[];
  account?: AccountObject;
  userActions: UserActionObject[];
}

// Note: This screen fetches data
export const MyAssetsScreen: React.FunctionComponent = () => {
  const dispatch = useDispatch();

  const { myAssets: assetsFromState, account, userActions } = useSelector<
    GlobalState,
    SelectedState
  >(state => {
    const { myAssets, accountInfo, userActions } = state;
    const { account } = accountInfo;
    const { list: myAssetsList } = myAssets;
    return { myAssets: myAssetsList, account, userActions };
  });

  useEffect(() => {
    const useEffectFunction = async (): Promise<void> => {
      if (account) {
        const { address } = account;
        const boughtAssets = _getBoughtAssetsFromState();

        const {
          assetsFromBlockchain,
          actionsFromBlockchain,
        } = await _getAssetsAndActionsFromBlockchain(address);

        const shouldUpdate = !listsAreEqual(assetsFromBlockchain, boughtAssets);
        if (shouldUpdate) {
          // TODO: Add some UI indication that something unexpected happened
          _logAssetsInconsistency(assetsFromBlockchain, boughtAssets);
          dispatch(removeFromMyAssetsList(boughtAssets));
          dispatch(appendToMyAssetsList(assetsFromBlockchain));
          dispatch(addUserAction(actionsFromBlockchain));
        }
      }
    };
    useEffectFunction();
  });

  const _getAssetsAndActionsFromBlockchain = async (
    address
  ): Promise<AssetsAndActionsObject> => {
    const assetsFromBlockchain = await _getAssetsFromBlockchain(address);
    let actionsFromBlockchain = {};
    assetsFromBlockchain.forEach((asset: AssetObject) => {
      const { actionId, id: assetId } = asset;

      const userAction = { [actionId]: { assetId, status: SUCCESSFUL } };
      actionsFromBlockchain = { ...actionsFromBlockchain, userAction };
    });
    assetsFromBlockchain.forEach(asset => delete asset.actionId);
    return { assetsFromBlockchain, actionsFromBlockchain };
  };

  const _getBoughtAssetsFromState = (): object[] => {
    const boughtAssets = assetsFromState.filter((asset): boolean => {
      const { id: assetId } = asset;
      const assetAction = userActions.find(
        action => action.assetId === assetId
      );
      return assetAction && assetAction.status === SUCCESSFUL;
    });
    return boughtAssets;
  };

  const _logAssetsInconsistency = (fromBlockchain, fromState): void => {
    const fromBlockchainIds = fromBlockchain.map(asset => asset.id);
    const fromStateIds = fromState.map(asset => asset.id);
    const addedIds = fromBlockchainIds.filter(id => !fromStateIds.includes(id));
    const removedIds = fromStateIds.filter(
      id => !fromBlockchainIds.includes(id)
    );
    logInfo(
      `Account's assets from blockchain aren't the same as the ones stored on the app.`
    );
    if (addedIds.length > 0)
      logInfo(`Some assets added to MyLandScreen. IDs: [${addedIds}]`);
    if (removedIds.length > 0)
      logWarn(`Some assets removed from MyLandScreen. IDs: [${removedIds}]`);
  };

  const _getAssetsFromBlockchain = async (address): Promise<AssetObject[]> => {
    const listOfPromises = await _getAssetsOf(address);
    const fromBlockchain = await Promise.all([...listOfPromises]);
    // Note: The UI shows land the user bought ordered by purchase date (desc)
    return fromBlockchain.reverse();
  };

  // Note: Returns a list of Promises
  const _getAssetsOf = async (address): Promise<Promise<AssetObject>[]> => {
    const decentralandUtils = new DecentralandUtils();
    const { getAssetsOf: getLandOf } = decentralandUtils;
    const contracts = getContracts();
    const { estateContract, landContract } = contracts;
    const listOfLand = await getLandOf(address);
    const assets = [];
    let land;

    for (land of listOfLand) {
      const generateAsset = async (land): Promise<object> => {
        const { id: assetId, nftAddress, transactionHash: actionId } = land;
        const asset = await generateAssetFromId(
          estateContract,
          landContract,
          assetId,
          nftAddress
        );
        return { ...asset, actionId };
      };

      const assetPromise = generateAsset(land);
      assets.push(assetPromise);
    }
    return assets;
  };

  const _renderItem = ({ item: assetAndUserAction }): JSX.Element => {
    const { asset } = assetAndUserAction;
    let { userAction } = assetAndUserAction;
    if (!userAction) userAction = {};
    return <MyAssetsListItem asset={asset} userAction={userAction} />;
  };

  return (
    <MyAssetsList
      myAssets={assetsFromState}
      userActions={userActions}
      renderItem={_renderItem}
    />
  );
};

export default MyAssetsScreen;
