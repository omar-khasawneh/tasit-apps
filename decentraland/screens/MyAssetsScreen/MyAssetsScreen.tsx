import React from "react";
import { connect } from "react-redux";
import {
  removeFromMyAssetsList,
  appendToMyAssetsList,
  addUserAction,
} from "../../redux/actions";
import MyAssetsList from "../../components/presentational/MyAssetsList";
import MyAssetsListItem from "../../components/presentational/MyAssetsListItem";
import { listsAreEqual, getContracts, logInfo, logWarn } from "../../helpers";
import { generateAssetFromId } from "../../helpers/decentraland";
import DecentralandUtils from "tasit-sdk/dist/helpers/DecentralandUtils";
import { SUCCESSFUL } from "../../constants/UserActionStatus";

interface AccountObject {
  address: string;
}

interface AssetObject {
  actionId: string;
  id: string;
}

interface AssetsAndActionsObject {
  assetsFromBlockchain: AssetObject[];
  actionsFromBlockchain: object[];
}

type MyAssetsScreenProps = {
  myAssets: any[];
  account?: AccountObject;
  userActions: any;
  removeFromMyAssetsList: (...args: any[]) => any;
  appendToMyAssetsList: (...args: any[]) => any;
  addUserAction: (...args: any[]) => any;
};

export class MyAssetsScreen extends React.Component<MyAssetsScreenProps, {}> {
  componentDidMount = async () => {
    const {
      account,
      removeFromMyAssetsList,
      appendToMyAssetsList,
      addUserAction,
    } = this.props;

    if (account) {
      const { address } = account;
      const boughtAssets = this._getBoughtAssetsFromState();

      const {
        assetsFromBlockchain,
        actionsFromBlockchain,
      } = await this._getAssetsAndActionsFromBlockchain(address);

      const shouldUpdate = !listsAreEqual(assetsFromBlockchain, boughtAssets);
      if (shouldUpdate) {
        // TODO: Add some UI indication that something unexpected happened
        this._logAssetsInconsistency(assetsFromBlockchain, boughtAssets);
        removeFromMyAssetsList(boughtAssets);
        appendToMyAssetsList(assetsFromBlockchain);
        addUserAction(actionsFromBlockchain);
      }
    }
  };

  _getAssetsAndActionsFromBlockchain = async (
    address
  ): Promise<AssetsAndActionsObject> => {
    const assetsFromBlockchain = await this._getAssetsFromBlockchain(address);
    let actionsFromBlockchain = {};
    assetsFromBlockchain.forEach((asset: AssetObject) => {
      const { actionId, id: assetId } = asset;

      const userAction = { [actionId]: { assetId, status: SUCCESSFUL } };
      actionsFromBlockchain = { ...actionsFromBlockchain, userAction };
    });
    assetsFromBlockchain.forEach(asset => delete asset.actionId);
    return { assetsFromBlockchain, actionsFromBlockchain };
  };

  _getBoughtAssetsFromState = (): object[] => {
    const { myAssets: assetsFromState, userActions } = this.props;
    const boughtAssets = assetsFromState.filter(asset => {
      const { id: assetId } = asset;
      const assetAction = userActions.find(
        action => action.assetId === assetId
      );
      return assetAction && assetAction.status === SUCCESSFUL;
    });
    return boughtAssets;
  };

  _logAssetsInconsistency = (fromBlockchain, fromState): void => {
    const fromBlockchainIds = fromBlockchain.map(asset => asset.id);
    const fromStateIds = fromState.map(asset => asset.id);
    const addedIds = fromBlockchainIds.filter(id => !fromStateIds.includes(id));
    const removedIds = fromStateIds.filter(
      id => !fromBlockchainIds.includes(id)
    );
    logInfo(
      `Account's assets from blockchain aren't the same as the ones stored on the app.`
    );
    if (addedIds.length > 0)
      logInfo(`Some assets added to MyLandScreen. IDs: [${addedIds}]`);
    if (removedIds.length > 0)
      logWarn(`Some assets removed from MyLandScreen. IDs: [${removedIds}]`);
  };

  _getAssetsFromBlockchain = async (address): Promise<AssetObject[]> => {
    const listOfPromises = await this._getAssetsOf(address);
    const fromBlockchain = await Promise.all([...listOfPromises]);
    // Note: The UI shows land the user bought ordered by purchase date (desc)
    return fromBlockchain.reverse();
  };

  // Note: Returns a list of Promises
  _getAssetsOf = async (address): Promise<Promise<any>[]> => {
    const decentralandUtils = new DecentralandUtils();
    const { getAssetsOf: getLandOf } = decentralandUtils;
    const contracts = getContracts();
    const { estateContract, landContract } = contracts;
    const listOfLand = await getLandOf(address);
    const assets = [];
    let land;

    for (land of listOfLand) {
      const generateAsset = async (land): Promise<object> => {
        const { id: assetId, nftAddress, transactionHash: actionId } = land;
        const asset = await generateAssetFromId(
          estateContract,
          landContract,
          assetId,
          nftAddress
        );
        return { ...asset, actionId };
      };

      const assetPromise = generateAsset(land);
      assets.push(assetPromise);
    }
    return assets;
  };

  _renderItem = ({ item: assetAndUserAction }): JSX.Element => {
    const { asset } = assetAndUserAction;
    let { userAction } = assetAndUserAction;
    if (!userAction) userAction = {};
    return <MyAssetsListItem asset={asset} userAction={userAction} />;
  };

  render(): JSX.Element {
    const { myAssets, userActions } = this.props;
    return (
      <MyAssetsList
        myAssets={myAssets}
        userActions={userActions}
        renderItem={this._renderItem}
      />
    );
  }
}
const mapStateToProps = (state): object => {
  const { myAssets, accountInfo, userActions } = state;
  const { account } = accountInfo;
  const { list: myAssetsList } = myAssets;
  return { myAssets: myAssetsList, account, userActions };
};
const mapDispatchToProps = {
  removeFromMyAssetsList,
  appendToMyAssetsList,
  addUserAction,
};
export default connect(
  mapStateToProps,
  mapDispatchToProps
)(MyAssetsScreen);
