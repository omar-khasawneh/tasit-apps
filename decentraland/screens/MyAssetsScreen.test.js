import React from "react";
import { shallow } from "enzyme";
import { MyAssetsScreen } from "./MyAssetsScreen";
import { parcel, parcelUserAction } from "@helpers/testHelpers";

describe("ListLandForSaleScreen", () => {
  describe("renders the component", () => {
    it("without assets", async () => {
      const myAssets = [];
      const userActions = {};
      expect(
        shallow(
          <MyAssetsScreen myAssets={myAssets} userActions={userActions} />
        )
      ).toMatchSnapshot();
    });
    it("with assets", async () => {
      const myAssets = [parcel];
      const userActions = { ...parcelUserAction };
      expect(
        shallow(
          <MyAssetsScreen myAssets={myAssets} userActions={userActions} />
        )
      ).toMatchSnapshot();
    });
  });
});
