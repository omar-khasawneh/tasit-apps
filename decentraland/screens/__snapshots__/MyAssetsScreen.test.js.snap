// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListLandForSaleScreen renders the component with assets 1`] = `
<MyAssetsList
  myAssets={
    Array [
      Object {
        "id": "0123456789",
        "img": "https://api.decentraland.org/v1/parcels/0/0/map.png",
        "name": "Sample Parcel",
        "type": "PARCEL",
      },
    ]
  }
  renderItem={[Function]}
  userActions={
    Array [
      Object {
        "actionId": "0x1234567890123456789012345678901234567890",
        "assetId": "0123456789",
        "status": "SUCCESSFUL",
      },
    ]
  }
/>
`;

exports[`ListLandForSaleScreen renders the component without assets 1`] = `
<MyAssetsList
  myAssets={Array []}
  renderItem={[Function]}
  userActions={Array []}
/>
`;
